text/0000-stdsimd.md

```diff
@@ -0,0 +1,85 @@
+- Feature Name: stdsimd_project_group
+- Start Date: 2020-08-28
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This is a project group RFC version of [`lang-team#29`].
+
+This RFC establishes a new project group, under the libs team, to produce a portable SIMD API in a new `rust-lang/stdsimd` repository, exposed through a new `std::simd` (and `core::simd`) module in the standard library in the same manner as [`stdarch`]. The output of this project group will be the finalization of [RFC 2948] and stabilization of `std::simd`.
+
+# Motivation
+[motivation]: #motivation
+
+The current stable `core::arch` module is described by [RFC 2325], which considers a portable API desirable but out-of-scope. The current [RFC 2948] provides a good motivation for this API. Various ecosystem implementations of portable SIMD have appeared over the years, including [`packed_simd`], and [`wide`], each taking a different set of trade-offs in implementation while retaining some similarities in their public API. The group will pull together a "blessed" implementation in the standard library with the explicit goal of stabilization for the [2021 edition].
+
+# Charter
+[charter]: #charter
+
+## Goals
+
+- Determine the shape of the portable SIMD API.
+- Get an unstable `std::simd` and `core::simd` API in the standard library. This may mean renaming `packed_simd` to `stdsimd` and working directly on it, or creating a new repository and pulling in chunks of code as needed.
```

Comment:
Shouldn't we evaluate options other than `packed_simd` and `stdsimd` too or try experimenting more? Or are these already good enough?

Reply:
This is definitely something to figure out when we kick off, and at the very least we should survey the various approaches out there. I’d love to take some time to dig into [`generic-simd`](https://docs.rs/generic-simd/0.1.0/generic_simd/index.html) properly.

In general I think we’d settled on a starting point based off `packed_simd`, without exposing a generic `Simd<T>` (so a public structure similar to `wide`), with a sprinkling of const generics for shuffles.

Reply:
> Shouldn't we evaluate options other than `packed_simd` and `stdsimd` too or try experimenting more? Or are these already good enough?

I think more experimentation carries a significant risk of not reaching the goal of getting to stable.

Notably, `simd` already existed before `packed_simd`, the initial design of `packed_simd` omitted something that `simd` already had (boolean/mask vectors) and ended up adding them back, so `packed_simd` ended up being like a more comprehensive version of `simd` with the internals rewritten.

That is, portable SIMD in Rust has already gone through the cycle of a change in the person working on it resulting in a rewrite that, while making things more comprehensive, on the high level ended up close to the old thing. While each such cycle might improved the design marginally, starting over means the distance to getting the stable gets reset to zero.

Reply:
Ah, so the goal here is more of stabilizing it rather than having more experimentation to improve the existing ones.

Reply:
I would also like to mention that even though I started from scratch, `generic-simd` ended up particularly similar to `packed_simd`, and most of the features would be applicable to an implementation based on `packed_simd`.

The major differences of `generic-simd` are:
* works on stable (likely irrelevant here)
* a little bit more comprehensive traits
* a different model for runtime feature selection (not sure if this is relevant)

Reply:
> a different model for runtime feature selection (not sure if this is relevant)

I'd say that runtime feature selection is a very important concern, as it's the only way to support more than just the lowest common denominator for your target platform.

The story for runtime feature detection in packed_simd is unclear to me, it doesn't immediatly stand out in the documentation whether it is really supported. But I'd think the finer details are a better fit for a zulip thread at some point in the future.

Reply:
A solution without runtime feature detection is a non-starter for me, so I agree.  `packed_simd` uses intrinsics so I believe it should "just work" with `#[target_feature]`, which is very convenient.  But I agree that this may be getting too much into details for now.

Reply:
I think run-time selection is scope creep. It also applies to other things already in the standard library: `count_ones` in particular. I think it should be developed separately and in a way that covers `count_ones`, too.

Reply:
It's my expectation that `packed_simd` already works with runtime selection, as @calebzulawksi says. `packed_simd` is using LLVM's primitives to implement SIMD support, so when it does code generation for a function tagged with `#[target_feature]`, then it should produce code for that target feature, just like LLVM does with anything else. And this would similarly work for things like `count_ones` as well, I imagine.

It should be easy to test anyway.

Reply:
2 things:

1) `wide` has *also* gone through iterations and arrived at a design that's "close enough" to `packed_simd` that people can jump between using the two crates via feature flag with only a small rename here and there. When adding in the `generic-simd` testimony above, it's safe to say that the broad level of the design is solid enough to begin, and any fine details can be solved during the project group.
2) I would like to confirm that `packed_simd` does not *on its own* provide any sort of runtime feature selection at all. However, with `#[target_feature(enable = "foo")]` you can generate functions that assume a given feature level greater than the program's default, and crates such as [multiversion](https://docs.rs/multiversion/) exist to ease the process of having a single function generated at more than one feature level and then making a selection of what inner function to call at runtime. Presuming that almost everything in `core::simd` will be marked as `#[inline]`, and thus codegen in the user's crate, this will all come together just fine.

And as less of a full point, is agree with @calebzualawski that we will likely want to be *more* trait oriented to allow for ease of user code being generic across lane count.

Reply:
So I believe we are looking for something that can have both with and without runtime selection? That sounds like it will add "what color is my function?" to me in case depending on how it was designed like the current async.

Reply:
No, `core::simd` will never look at available features. It just calls intrinsics, and llvm does the conversion, and based on codegen settings in the calling code you get what you get.

Reply:
Note that LLVM can't inline functions that have incompatible `target-features` attributes ([source](https://github.com/llvm/llvm-project/blob/3e4e0fb2435544acadf3614d3cd7b5f0f8fdfda2/llvm/include/llvm/Analysis/TargetTransformInfoImpl.h#L603)). This means that `core::simd` will only use hardware SIMD instructions if SIMD features are enabled globally.

Reply:
I was under the impression that `#[target_feature(enable = "avx")]` would also have the same effect (but make the function unsafe to call).

Reply:
Yes, but the question is, what attributes are you going to put on the function in `core::simd`?

Reply:
Actually looking at the LLVM code some more, it seems you can bypass this check by using `#[inline(always)]`. Please ignore my previous comment.

Reply:
Just `#[inline]` goes on the core::simd function, and then the caller can choose to target_feature(enable) in a function that uses core::simd.

According to the docs for `is_x86_feature_detected!`, that alone should do it, without core::simd needing to be target feature aware at all.

Reply:
You need to use `#[inline(always)]` on the `core::simd` function otherwise LLVM will refuse to inline functions with mismatching target features. But otherwise you are correct.

Reply:
How about runtime selection? Is rust going to provide runtime selection for SIMD which may not be zero-cost? Or maybe users can opt-in for it?

Reply:
That is what the `multiversion` crate (or similar) would do.

Actually putting that sort of mechanic in the standard library is probably a whole separate project.